qemu/tests/block_hotplug.py

```diff
@@ -88,153 +81,85 @@ def unplug_device(vm, get_disk_cmd, device):
-                                                            disk_letter[num])
-                    test_cmd = utils_misc.set_winutils_letter(session, test_cmd)
-                else:
-                    test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
-
-                status, output = session.cmd_status_output(test_cmd,
-                                                           timeout=disk_op_timeout)
-                if status:
-                    test.fail("Check for block device failed "
-                              "after hotplug, Output: %r" % output)
-            session.close()
-
-        sub_type = params.get("sub_type_after_plug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "after hotplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
-            if sub_type == "shutdown" and vm.is_dead():
-                return
-
-        sub_type = params.get("sub_type_before_unplug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "before unplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
+        plug_tag = "before_unplug"
+        run_sub_test(params, plug_tag)
 
         error_context.context("Unplug block device (iteration %d)" % iteration,
                               logging.info)
-        for controller in controller_list:
-            controller_id = controller.get_param("id")
-            for device in controller_device_dict.get(controller_id, []):
-                unplug_status = unplug_device(vm, get_disk_cmd, device)
-                if not unplug_status:
-                    test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
-                device_list.remove(device)
-            controller.unplug(vm.monitor)
-        for device in device_list:
-            unplug_status = unplug_device(vm, get_disk_cmd, device)
-            if not unplug_status:
-                test.fail("Failed to unplug disks '%s'" % device.get_param("id"))
-
-        sub_type = params.get("sub_type_after_unplug")
-        if sub_type:
-            error_context.context(context_msg % (sub_type, "after unplug"),
-                                  logging.info)
-            utils_test.run_virt_sub_test(test, params, env, sub_type)
+        disks_before_unplug = find_disk(vm, get_disk_cmd)
+        for device in reversed(device_list):
+            ret = vm.devices.simple_unplug(device, vm.monitor)
+            if ret[1] is False:
+                test.fail("Failed to unplug device '%s'."
+                          "Ouptut:\n%s" % (dev, ret[0]))
+
+        unplug_disks = utils_misc.wait_for(lambda: get_new_disk(find_disk(vm, get_disk_cmd),
+                                           disks_before_unplug), pause)
+        if not unplug_disks:
```

Comment:
We should check that all the target disks has been unplugged in guest, so here need to check the number.